crates/iota-indexer/src/indexer_reader.rs

```diff
@@ -712,10 +708,65 @@ impl IndexerReader {
                 )
                 // we filter the `tx_global_order` table because it is indexed by digest,
                 // optimistic_transactions table is not
-                .filter(tx_global_order::tx_digest.eq_any(missing_digests))
+                .filter(tx_global_order::tx_digest.eq_any(digests))
                 .select(OptimisticTransaction::as_select())
                 .load::<OptimisticTransaction>(conn)
-        })?;
+        })
+    }
+
+    async fn get_single_transaction(
+        &self,
+        digest: TransactionDigest,
+    ) -> IndexerResult<Option<StoredTransaction>> {
+        let digest_bytes = digest.inner().to_vec();
+        let checkpointed_tx_future = {
+            let this = self.clone();
+            let digest_bytes = digest_bytes.clone();
+            tokio::task::spawn_blocking(move || {
+                this.get_checkpointed_transactions(&[digest_bytes])
+                    .map(|mut txs| txs.pop())
+            })
+        };
+        let optimistic_tx_future = {
+            let this = self.clone();
+            tokio::task::spawn_blocking(move || {
+                this.get_optimistic_transactions(&[digest_bytes])
+                    .map(|mut txs| txs.pop().map(Into::into))
+            })
+        };
+
+        let mut tx_futures = FuturesUnordered::new();
+        tx_futures.push(checkpointed_tx_future);
+        tx_futures.push(optimistic_tx_future);
```

Comment:
nit: We could use `FuturesUnordered::from_iter` for a more concise initialization:

```suggestion
        let mut tx_futures =
            FuturesUnordered::from_iter([checkpointed_tx_future, optimistic_tx_future]);
```